go/vt/vtgate/planbuilder/operators/horizon_planning.go

```diff
@@ -97,118 +200,60 @@ func planHorizon(ctx *plancontext.PlanningContext, in *Horizon, isRoot bool) (op
 	}
 
 	needsOrdering := len(qp.OrderExprs) > 0
-	canShortcut := isRoute && sel.Having == nil && !needsOrdering
-	_, isDerived := in.Source.(*Derived)
-
-	switch {
-	case qp.NeedsAggregation() || sel.Having != nil || sel.Limit != nil || isDerived || needsOrdering || qp.NeedsDistinct():
-		return nil, errNotHorizonPlanned
-	case canShortcut:
-		return planSingleRoute(rb, in)
-	default:
-		return pushProjections(ctx, qp, in.Source, isRoot)
+	canPushDown := isRoute && sel.Having == nil && !needsOrdering && !qp.NeedsAggregation() && !sel.Distinct && sel.Limit == nil
+
+	if canPushDown {
+		return rewrite.Swap(in, rb)
 	}
+
+	return expandHorizon(qp, in)
+}
+
+// horizonLike should be removed. we should use Horizon for both these cases
+type horizonLike interface {
+	ops.Operator
+	selectStatement() sqlparser.SelectStatement
+	src() ops.Operator
 }
```

Comment:
`horizonLike` should be removed from the codebase before we merge this or from the operator tree?

Reply:
Seems like https://github.com/vitessio/vitess/pull/12887 is closed and will be ported over to this Pull Request.

Reply:
yeah, and now we regretted this decision. we still want to remove derived, but not now. we'll do it after we are done with porting over the horizon planning to operators